src/runtime/undefined.go

```diff
@@ -1,10 +1,294 @@
 package runtime
 
-type Undefined struct{}
+import (
+	"fmt"
+	"github.com/gojinja/gojinja/src/errors"
+	"github.com/gojinja/gojinja/src/utils"
+	"log"
+	"reflect"
+	"strconv"
+)
 
-type IUndefined interface{}
+type Undefined struct {
+	hint   *string
+	obj    any
+	name   *string
+	exc    func(msg string) error
+	logger *log.Logger
+}
+
+type StrictUndefined struct {
+	Undefined
+}
+
+type DebugUndefined struct {
+	Undefined
+}
+
+type ChainableUndefined struct {
+	Undefined
+}
+
+func (Undefined) Undefined() {}
+
+type IUndefined interface {
+	Undefined()
+}
+
+func NewUndefined(hint *string, obj any, name *string, exc func(msg string) error, logger *log.Logger) Undefined {
+	if exc == nil {
+		exc = errors.TemplateError
+	}
+	return Undefined{hint, obj, name, exc, logger}
+}
+
+func NewStrictUndefined(hint *string, obj any, name *string, exc func(msg string) error, logger *log.Logger) StrictUndefined {
+	return StrictUndefined{NewUndefined(hint, obj, name, exc, logger)}
+}
+
+func NewChainableUndefined(hint *string, obj any, name *string, exc func(msg string) error, logger *log.Logger) ChainableUndefined {
+	return ChainableUndefined{NewUndefined(hint, obj, name, exc, logger)}
+}
+
+func NewDebugUndefined(hint *string, obj any, name *string, exc func(msg string) error, logger *log.Logger) StrictUndefined {
+	return StrictUndefined{NewUndefined(hint, obj, name, exc, logger)}
+}
+
+func (u Undefined) undefinedMessage() string {
+	if u.hint != nil {
+		return *u.hint
+	}
+	if _, ok := u.obj.(utils.Missing); ok {
+		return fmt.Sprintf("'%s' is undefined", *u.name)
+	}
+	// Following if is a rewrite of python code below. I don't undeestand neither the message nor the logic, as undefined_name ought to be Optional[string].
```

Comment:
Let's hope it's a mistake in jinja and that undefined_name shouldn't in fact be `Any`